subxt/src/backend/rpc/reconnecting_rpc_client/mod.rs

```diff
@@ -0,0 +1,630 @@
+            #[cfg(feature = "native")]
+            headers: self.headers,
+            max_redirections: self.max_redirections,
+            max_log_len: self.max_log_len,
+            id_kind: self.id_kind,
+            max_concurrent_requests: self.max_concurrent_requests,
+            request_timeout: self.request_timeout,
+            connection_timeout: self.connection_timeout,
+        }
+    }
+
+    #[cfg(feature = "native")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "native")))]
+    /// Configure the WebSocket ping/pong interval.
+    ///
+    /// Default: 30 seconds.
+    pub fn enable_ws_ping(mut self, ping_config: PingConfig) -> Self {
+        self.ping_config = Some(ping_config);
+        self
+    }
+
+    #[cfg(feature = "native")]
+    #[cfg_attr(docsrs, doc(cfg(feature = "native")))]
+    /// Disable WebSocket ping/pongs.
+    ///
+    /// Default: 30 seconds.
+    pub fn disable_ws_ping(mut self) -> Self {
+        self.ping_config = None;
+        self
+    }
+
+    /// Build and connect to the target.
+    pub async fn build(self, url: String) -> Result<Client, RpcError> {
+        let (tx, rx) = mpsc::unbounded_channel();
+        let client = Retry::new(self.retry_policy.clone(), || {
+            platform::ws_client(url.as_ref(), &self)
+        })
+        .await?;
+        let (reconn_tx, reconn_rx) = reconnect_channel();
+
+        platform::spawn(background_task(client, rx, url, reconn_tx, self));
+
+        Ok(Client {
+            tx,
+            reconnect: reconn_rx,
+        })
+    }
+}
+
+impl Client {
+    /// Perform a JSON-RPC method call.
+    pub async fn request(
+        &self,
+        method: String,
+        params: Option<Box<RawValue>>,
+    ) -> Result<Box<RawValue>, Error> {
+        let (tx, rx) = oneshot::channel();
+        self.tx
+            .send(Op::Call {
+                method,
+                params: RpcParams(params),
+                send_back: tx,
+            })
+            .map_err(|_| Error::Dropped)?;
+
+        rx.await.map_err(|_| Error::Dropped)?
+    }
+
+    /// Perform a JSON-RPC subscription.
+    pub async fn subscribe(
+        &self,
+        subscribe_method: String,
+        params: Option<Box<RawValue>>,
+        unsubscribe_method: String,
+    ) -> Result<Subscription, Error> {
+        let (tx, rx) = oneshot::channel();
+        self.tx
+            .send(Op::Subscription {
+                subscribe_method,
+                params: RpcParams::new(params),
+                unsubscribe_method,
+                send_back: tx,
+            })
+            .map_err(|_| Error::Dropped)?;
+        rx.await.map_err(|_| Error::Dropped)?
+    }
+
+    /// A future that resolves when the client has initiated a reconnection.
+    /// This method returns another future that resolves when the client has reconnected.
+    ///
+    /// This may be called multiple times.
+    pub async fn reconnect_initiated(&self) -> impl Future<Output = ()> + '_ {
```

Comment:
I have no good reason why to expose this but I thought it was quite useful to know when the client reconnects.
Otherwise it's only possible to get this information by adding logs for it..

Sure, a stream would work perhaps a little nicer API to use

Reply:
but we can remove these APIs if you think it's leaky

Reply:
I think it's just one of those things where it's easier to add an API later if somebody wants it than it is to take it away once it's already there, so I'm normally more cautious adding things unless I can think of good uses! But yeah up to you; I'd just make it return a Stream if we kept it :)

Reply:
Removed now